# Import Libraries

In [119]:
import numpy as np
import pandas as pd
import torch
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import LabelEncoder

In [120]:
df = pd.read_csv('https://raw.githubusercontent.com/gscdit/Breast-Cancer-Detection/refs/heads/master/data.csv')
df.head()

,id,diagnosis,radius_mean,texture_mean,perimeter_mean,area_mean,smoothness_mean,compactness_mean,concavity_mean,concave points_mean,...,texture_worst,perimeter_worst,area_worst,smoothness_worst,compactness_worst,concavity_worst,concave points_worst,symmetry_worst,fractal_dimension_worst,Unnamed: 32
0,842302,M,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,...,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890,NaN
1,842517,M,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,...,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902,NaN
2,84300903,M,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,...,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758,NaN
3,84348301,M,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,...,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300,NaN
4,84358402,M,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,...,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678,NaN


In [121]:
df.shape

(569, 33)

In [122]:
df.drop(columns=['id', 'Unnamed: 32'], inplace= True)

In [123]:
df.head()

,diagnosis,radius_mean,texture_mean,perimeter_mean,area_mean,smoothness_mean,compactness_mean,concavity_mean,concave points_mean,symmetry_mean,...,radius_worst,texture_worst,perimeter_worst,area_worst,smoothness_worst,compactness_worst,concavity_worst,concave points_worst,symmetry_worst,fractal_dimension_worst
0,M,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,0.2419,...,25.38,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890
1,M,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,0.1812,...,24.99,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902
2,M,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,0.2069,...,23.57,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758
3,M,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,0.2597,...,14.91,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300
4,M,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,0.1809,...,22.54,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678


# Train test split

In [124]:
X_train, X_test, y_train, y_test = train_test_split(df.iloc[:, 1:], df.iloc[:, 0], test_size=0.2)


In [125]:
df['diagnosis'].unique()

array(['M', 'B'], dtype=object)

## Why only transform(X_test)?

- This is the most important part.

- You do not want to calculate a new mean/std from the test data.

Suppose:

In [126]:
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [127]:
encoder = LabelEncoder()
y_train = encoder.fit_transform(y_train)
y_test = encoder.transform(y_test)

In [128]:
df.head()

,diagnosis,radius_mean,texture_mean,perimeter_mean,area_mean,smoothness_mean,compactness_mean,concavity_mean,concave points_mean,symmetry_mean,...,radius_worst,texture_worst,perimeter_worst,area_worst,smoothness_worst,compactness_worst,concavity_worst,concave points_worst,symmetry_worst,fractal_dimension_worst
0,M,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,0.2419,...,25.38,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890
1,M,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,0.1812,...,24.99,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902
2,M,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,0.2069,...,23.57,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758
3,M,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,0.2597,...,14.91,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300
4,M,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,0.1809,...,22.54,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678


In [129]:
len(df.columns)

31

In [130]:
X_train

array([[-0.52348759, -0.36210776, -0.5549868 , ..., -0.81577297,
        -1.05631858, -0.53197725],
       [ 0.46315309, -0.00748518,  0.37334575, ..., -0.40672863,
        -0.32830457, -0.83252217],
       [-0.028729  , -0.64304253, -0.08768568, ..., -0.7639952 ,
        -0.33469065, -0.35505684],
       ...,
       [-0.31350284, -0.70982211, -0.3568645 , ..., -0.45744038,
         0.21610942, -0.7665891 ],
       [ 2.30411528,  0.57280631,  2.34620916, ...,  1.99210473,
        -0.23570629,  0.12625459],
       [ 0.21289729,  1.3833722 ,  0.1518166 , ..., -0.04397971,
        -0.26604021, -0.762743  ]])

In [131]:
y_train

array([0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0,
       1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1,
       0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 0,
       1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1,
       1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 1, 1,
       0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1,
       0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0,
       1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0,
       0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0,
       1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1,
       0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0,
       1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0,

In [132]:
X_train

array([[-0.52348759, -0.36210776, -0.5549868 , ..., -0.81577297,
        -1.05631858, -0.53197725],
       [ 0.46315309, -0.00748518,  0.37334575, ..., -0.40672863,
        -0.32830457, -0.83252217],
       [-0.028729  , -0.64304253, -0.08768568, ..., -0.7639952 ,
        -0.33469065, -0.35505684],
       ...,
       [-0.31350284, -0.70982211, -0.3568645 , ..., -0.45744038,
         0.21610942, -0.7665891 ],
       [ 2.30411528,  0.57280631,  2.34620916, ...,  1.99210473,
        -0.23570629,  0.12625459],
       [ 0.21289729,  1.3833722 ,  0.1518166 , ..., -0.04397971,
        -0.26604021, -0.762743  ]])

In [133]:
type(y_train)

numpy.ndarray

In [134]:
type(y_test)

numpy.ndarray

# Numpy Array to tensors

In [135]:
# Numpy arrays → PyTorch tensors (clean & recommended way)
# -------------------------------------------------------
# torch.from_numpy() shares memory with the numpy array.
# .float() converts to torch.float32 (required by most layers).

X_train_tensor = torch.from_numpy(X_train).float()
X_test_tensor  = torch.from_numpy(X_test).float()

y_train_tensor = torch.from_numpy(y_train).float()
y_test_tensor  = torch.from_numpy(y_test).float()

print("X_train_tensor shape:", X_train_tensor.shape)
print("X_test_tensor shape :", X_test_tensor.shape)
print("y_train_tensor shape:", y_train_tensor.shape)
print("y_test_tensor shape :", y_test_tensor.shape)


X_train_tensor shape: torch.Size([455, 30])
X_test_tensor shape : torch.Size([114, 30])
y_train_tensor shape: torch.Size([455])
y_test_tensor shape : torch.Size([114])


### Fix Applied (important)

The previous version had a bug:

```python
X_train_tensor = torch.from_numpy(X_train)
X_train_tensor = torch.tensor(X_test_tensor, ...)   # ← overwrote train with test data!
```

This caused the shape mismatch error:
`target size (455, 1)` vs `input size (114, 1)`.

Now both tensors are created correctly and converted to `float32` in one clean step.

In [136]:
type(X_train_tensor)

torch.Tensor

# Defining the Model

In [137]:
X_train.shape[1]

30

In [138]:
import torch.nn as nn

In [139]:
class MySimpleNN(nn.Module):
    # MySimpleNN is our own neural network class.
    # nn.Module is the base class provided by PyTorch.
    # By inheriting from nn.Module, PyTorch knows this is a neural network.

    def __init__(self, num_features):
        # __init__ runs when we create the model.
        #
        # num_features = number of input features.
        #
        # Example:
        # If each person has:
        # age, income, height, weight, experience
        #
        # Then:
        # num_features = 5

        super().__init__()
        # Initialize nn.Module.
        #
        # Think of this as:
        # "PyTorch, please initialize all the neural-network
        # functionality for my model."

        self.linear = nn.Linear(num_features, 1)
        # Create a Linear layer.
        #
        # It takes:
        #     num_features inputs
        #
        # and produces:
        #     1 output
        #
        # Mathematically:
        #
        # output = XW + b
        #
        # If num_features = 5:
        #
        #     5 inputs ---> Linear ---> 1 output
        #
        # The Linear layer automatically creates:
        #     weights (W)
        #     bias (b)

        self.sigmoid = nn.Sigmoid()
        # Create a Sigmoid activation function.
        #
        # Sigmoid converts the output into a value between 0 and 1.
        #
        # Example:
        #     3.5  ---> 0.9707
        #    -2.0  ---> 0.1192
        #     0.0  ---> 0.5000
        #
        # This is commonly used for binary classification.

    def forward(self, features):
        # forward() describes how data moves through the network.
        #
        # features = input data.
        #
        # Example:
        # features =
        # [age, income, height, weight, experience]

        out = self.linear(features)
        # Send the input through the Linear layer.
        #
        # features
        #    ↓
        # Linear
        #    ↓
        # out
        #
        # Mathematically:
        # out = features × weights + bias

        out = self.sigmoid(out)
        # Now send the Linear output through Sigmoid.
        #
        # Linear output:
        #     -2.5
        #
        # Sigmoid output:
        #     0.0759
        #
        # So:
        #
        # Linear → Sigmoid

        return out
        # Return the final prediction.


In [140]:
torch.rand(30,1)

tensor([[7.8994e-04],
        [1.9138e-01],
        [1.6623e-01],
        [5.2962e-03],
        [5.7395e-01],
        [3.8892e-01],
        [9.9518e-01],
        [9.5172e-01],
        [3.3343e-01],
        [4.0718e-01],
        [4.6092e-01],
        [7.4765e-01],
        [2.2340e-01],
        [6.9238e-01],
        [6.9472e-01],
        [6.1230e-01],
        [8.5603e-02],
        [4.0542e-01],
        [3.7913e-01],
        [1.0744e-01],
        [6.0925e-01],
        [7.6641e-01],
        [2.8530e-01],
        [7.2678e-01],
        [9.9615e-01],
        [7.3813e-01],
        [3.0904e-01],
        [3.9118e-01],
        [1.8602e-01],
        [8.3310e-01]])

# Parameters

In [141]:
X_train_tensor.shape[1]

30

In [142]:
learning_rate = 0.1
epochs = 25

In [143]:
loss_function = nn.BCELoss()

In [144]:
type(loss_function)

torch.nn.modules.loss.BCELoss

In [145]:
model = MySimpleNN(X_train_tensor.shape[1])


In [146]:
model.parameters()

<generator object Module.parameters at 0x7df8c0643920>

In [147]:
X_test_tensor.dtype

torch.float32

In [148]:
model.linear.weight.dtype

torch.float32

In [149]:
y_train_tensor.dtype

torch.float32

In [150]:
model.parameters()

<generator object Module.parameters at 0x7df8c0643bc0>

# Training Pipeline

                TRAINING
                   │
                   ↓
          ┌─────────────────┐
          │  X_train_tensor │
          └────────┬────────┘
                   │
                   ↓
              ┌─────────┐
              │  MODEL  │
              └────┬────┘
                   │
                   ↓
                y_pred
                   │
                   │ compare
                   ↓
              ┌─────────┐
              │  LOSS   │ ← y_train
              └────┬────┘
                   │
                   ↓
          loss.backward()
                   │
                   ↓
              Gradients
                   │
                   ↓
          optimizer.step()
                   │
                   ↓
            Update weights
                   │
                   ↓
                Repeat

In [151]:
# Create model
model = MySimpleNN(X_train_tensor.shape[1])

# Create optimizer
optimizer = torch.optim.SGD(
    model.parameters(),
    lr=learning_rate
)

# Training loop
for epoch in range(epochs):

    # -------------------------
    # 1. Forward Pass
    # -------------------------
    # Give training features to the model.
    #
    # X_train_tensor:
    # [number_of_samples, number_of_features]
    #
    # Model:
    # Linear → Sigmoid
    #
    # Result:
    # y_pred = predicted probabilities

    y_pred = model(X_train_tensor)


    # -------------------------
    # 2. Calculate Loss
    # -------------------------
    # Compare:
    #
    # y_pred  → model prediction
    # y_train → actual answer
    #
    # Both MUST have the same shape.

    loss = loss_function(
        y_pred,
        y_train_tensor.reshape(-1, 1)
    )


    # -------------------------
    # 3. Zero Gradients
    # -------------------------
    # Remove gradients from the previous iteration.

    optimizer.zero_grad()


    # -------------------------
    # 4. Backward Pass
    # -------------------------
    # Calculate gradients of the loss
    # with respect to model parameters.

    loss.backward()


    # -------------------------
    # 5. Update Parameters
    # -------------------------
    # SGD uses the gradients to update
    # weights and biases.

    optimizer.step()


    # -------------------------
    # 6. Print Loss
    # -------------------------

    print(f"Epoch {epoch + 1}, Loss = {loss.item()}")

Epoch 1, Loss = 0.5913804769515991
Epoch 2, Loss = 0.47829103469848633
Epoch 3, Loss = 0.41227051615715027
Epoch 4, Loss = 0.36834946274757385
Epoch 5, Loss = 0.3365417420864105
Epoch 6, Loss = 0.312160462141037
Epoch 7, Loss = 0.2927047610282898
Epoch 8, Loss = 0.27670931816101074
Epoch 9, Loss = 0.26325318217277527
Epoch 10, Loss = 0.2517254650592804
Epoch 11, Loss = 0.24170321226119995
Epoch 12, Loss = 0.23288315534591675
Epoch 13, Loss = 0.22504156827926636
Epoch 14, Loss = 0.21800939738750458
Epoch 15, Loss = 0.2116561084985733
Epoch 16, Loss = 0.20587913691997528
Epoch 17, Loss = 0.20059674978256226
Epoch 18, Loss = 0.19574271142482758
Epoch 19, Loss = 0.19126291573047638
Epoch 20, Loss = 0.18711240589618683
Epoch 21, Loss = 0.1832536906003952
Epoch 22, Loss = 0.17965508997440338
Epoch 23, Loss = 0.1762896478176117
Epoch 24, Loss = 0.17313416302204132
Epoch 25, Loss = 0.1701687127351761


# Model Evaluation

In [152]:
with torch.no_grad():
  y_pred = model.forward(X_test_tensor)
  y_pred= (y_pred > 0.9).float()
  accuracy = (y_pred == y_test_tensor).float().mean()
  print(f" Accuracy = {accuracy.item()}")

 Accuracy = 0.53847336769104


# Summary

In [153]:
X_train_tensor.shape

torch.Size([455, 30])

In [154]:
!pip install torchinfo

In [155]:
from torchinfo import summary

In [156]:
summary(model, input_size = (455,30))

Layer (type:depth-idx)                   Output Shape              Param #
MySimpleNN                               [455, 1]                  --
├─Linear: 1-1                            [455, 1]                  31
├─Sigmoid: 1-2                           [455, 1]                  --
Total params: 31
Trainable params: 31
Non-trainable params: 0
Total mult-adds (Units.MEGABYTES): 0.01
Input size (MB): 0.05
Forward/backward pass size (MB): 0.00
Params size (MB): 0.00
Estimated Total Size (MB): 0.06